### PR TITLE
feat: support jest v27 in vue3-jest

### PR DESCRIPTION
**Pull request by cexbrayat**
And also babel-jest and ts-jest v27.

The PR is currently failing because of a change/regression in ts-jest (update: fixed in ts-jest v27.0.1)

TODO:
- [x] wait for https://github.com/kulshekhar/ts-jest/pull/2628 to be merged and released

### CODE REVIEW

**Comment by jurijzahn8019** [comment]
Hi,

https://github.com/kulshekhar/ts-jest/pull/2628 was merged

And ts-jest 27 was released :+1:

**Comment by cexbrayat** [comment]
ts-jest v27.0.1 was released including the fix, the PR is now green.

File: lib/transformers/typescript.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 const ensureRequire = require('../ensure-require')
-const babelJest = require('babel-jest')
+const babelJest = require('babel-jest').default
```
Comment (lmiller1990) [inline lib/transformers/typescript.js:2]: Has jest changed how it processes modules with default exports? I guess this is the breaking change?

File: lib/utils.js [hunk 2]
```diff
@@ -49,7 +49,11 @@ const transformContent = function transformContent(
 
 const getVueJestConfig = function getVueJestConfig(jestConfig) {
   return (
-    (jestConfig && jestConfig.globals && jestConfig.globals['vue-jest']) || {}
+    (jestConfig &&
```
Comment (lmiller1990) [inline lib/utils.js:52]: Right, this will also be a breaking change.

**Review by lmiller1990 — COMMENTED** [review]
I see we have some breaking changes (did not go over Jest 27 release notes yet).

I wonder if we can just do something simple like:

```
const jestVersion = require('jest/package.json').version

if (jestVersion.includes('26')) {
  // v26 specific things
}
```

Versioning for vue-jest has been a problem for a while. We have several moving targets - Vue 2/3, and Jest 26/27. It's not entirely clear on the best way to do this. What do you think?

**Comment by nogic1008** [comment]
vue-jest@v5 (for Vue 3, Jest 24~26) is still in alpha, so we can include breaking changes even if following semver.
But, v4 was public released.
If we follow semver, we should increment major version.

How about separating `next` branch into another package (such as `vue3-jest`)?

**Comment by cexbrayat** [comment]
I think a clean way would be to do the same than ts-jest and babel-jest for the versioning: use the jest version as the major one. That would mean releasing vue-jest v27, compatible with jest v27.

If jest 27 support for Vue 2 is important, maybe having another package could avoid the confusion (vue2-jest v27 and vue3-jest v27).

This is definitely a big change but the current versioning is confusing as it is, and I guess more and more projects starting now will bet on Vue 3 and Jest 27. As a developer, it would be clear that using vue-jest v27 (or vue3-jest v27)  is what you need.

**Comment by lmiller1990** [comment]
I like the idea to match the Jest versioning. The problem historically was how to we handle Vue 2? And eventually, Vue 3?

A separate package would likely be the easiest solution. In fact, two separate packages, even? Should we deprecate `vue-jest` entirely, in favor of:

- `vue2-jest26`
- `vue2-jest27`
- `vue3-jest27`

etc?

Probably a good opportunity to introduce semantic release, too. Currently I do releases manually, which is not very scalable.

**Comment by wolfgangwalther** [comment]
> * `vue2-jest26`
> 
> * `vue2-jest27`
> 
> * `vue3-jest27`

That's one package for each jest major? Or rather `vue2-jest@26.x.x`, `vue2-jest@27.x.x`, `vue3-jest@27.x.x`? That would make the most sense, I guess.

**Comment by yyx990803** [comment]
I think it makes sense for `vue-jest` to match Jest versions and support the current latest Vue major (where Vue 3 is becoming the default soon). A separate package to support older versions of Vue - i.e.

- `vue-jest@27` -> Vue 3 + Jest 27
- `vue2-jest@27` -> Vue 2 + Jest 27

**Comment by lmiller1990** [comment]
@wolfgangwalther it's not necessarily per major, it can be minor (or) patch. Take a look at [ts-jest](https://github.com/kulshekhar/ts-jest/releases). They basically just do a release when it makes sense, but generally try to be up to date where possible and practical.

@yyx990803 this seems like a pretty reasonable approach. If (when) we have Vue 4, we can create `vue3-jest@x` and `vue-jest` would then default to Vue 4 (thinking really far ahead here, but good to consider these things).

In this case, we can start preparing this now, and make `vue-jest` default to Vue 3 + Jest 27 when Vue 3 becomes the default on npm. I think a good short term goal is:

- `vue-jest@27` -> Vue 3 / Jest 27 (this PR)
- `vue-jest@26` -> Vue 3 / Jest 26 (already have this)
- `vue2-jest@27` -> Vue 2 / Jest 27 (people will still need this, we will need to back port this PR - not hard)
- `vue2-jest@26` -> Vue 2 / Jest 26 (already exists)

Anyone want to work on this? If not, happy to pick it up, but I won't be able to until around the middle of this month (so, 2 weeks-ish from this post).

**Comment by Kocal** [comment]
Thanks for your work @cexbrayat! :heart:

**Comment by nogic1008** [comment]
Is it difficult to make this repository to monorepo?
Like this
- `/`
  - `/packages`
    - `/vue2` - npm package for Vue 2 (named `vue2-jest`?)
      - `/src`
      - `/test`
      - …
    - `/vue3` - npm package for Vue 3 (named `vue-jest` or `vue3-jest`?)
      - `/src`
      - `/test`
      - …
    - `/vue-next` - npm package for Vue next version
      - `/src`
      - `/test`
      - …
    - `/common` - common logic (if available)
      - `/src`
      - `/test`
      - …

**Comment by lmiller1990** [comment]
@nogic1008 that's probably a good idea. Then we can manage everything here (and share code).

I have not set up a monorepo before - is this something you have experience with and would be interested in doing?

**Comment by nogic1008** [comment]
@lmiller1990 OK.
I'll make a monorepo sample in my repository. After that, consider the migration plan.

**Comment by lmiller1990** [comment]
@nogic1008, ok thanks! You should have access to push branches here too; you could just make a branch in this repo if you like.

If you don't have push access, let me know, we will get it set up. We should do the same for @cexbrayat, since you guys are contributing a lot.

**Comment by Tofandel** [comment]
I like the idea to have version like
`vue-jest@27` -> Vue 3 + Jest 27
`vue2-jest@27` -> Vue 2 + Jest 27

Where the default `vue-jest` always tracks the latest vue version, so if there is a vue v4, this would be in `vue-jest` and a new `vue3-jest` would have to be created but that also means the package would need to be changed for users of the lib

So I'm actually more enclined to have both
`vue-jest@27` -> Vue 3 + Jest 27
`vue3-jest@27` -> Vue 3 + Jest 27
`vue2-jest@27` -> Vue 2 + Jest 27

Where the first one is a mirror of the latest vue, and if that's too much to maintain or bad to have npm mirrors for whatever reason then I'd be more inclined for only:
`vue3-jest@27` -> Vue 3 + Jest 27
`vue2-jest@27` -> Vue 2 + Jest 27

And deprecate 
`vue-jest`

**Comment by nogic1008** [comment]
Until the package naming and versioning policy is decided, this PR cannot be merged.
I opened the issue with #355.
If you have any idea, please tell me there.

**Comment by lmiller1990** [comment]
I think it's probably okay to leave this PR open to give some visibility to the required changes 🤔 

I can give you some feedback soon! We will want to include the changes in this PR eventually.

**Comment by lmiller1990** [comment]
Hello all, quick update.

This will be merged and released after this one: https://github.com/vuejs/vue-jest/pull/361 where we move to a monorepo, allowing us to simultaneously support multiple versions of Jest. That should be soon (next week or so).

**Comment by lmiller1990** [comment]
New update, we have now got master and next on a monorepo.

Next steps here: https://github.com/vuejs/vue-jest/issues/355#issuecomment-864778815.

I think we can finish this all before the end of June.

**Comment by cexbrayat** [comment]
@lmiller1990 do you need me to rebase this PR now?

**Comment by lmiller1990** [comment]
Yeah that would be great @cexbrayat - add a new package for the Jest 27 integration.

I will figure out the release process on the weekend. Let's make sure goes smoothly!

**Comment by lmiller1990** [comment]
Sorry all - I got caught up in many other things. 

@cexbrayat do you have time to remake this PR forked from master? Master will now be supporting Jest 27 by default. [See this comment](https://github.com/vuejs/vue-jest/issues/355#issuecomment-864778815).

We now have a 26.x branch - that will be to Jest 26 support (it's currently even with master).

Once we have Jest 27 ready to go, we can go ahead and start moving forward with releases matching Jest, and not using semantic versioning anymore.

**Comment by cexbrayat** [comment]
@lmiller1990 Sure, I'll take a look in the next few days 👍

**Comment by cexbrayat** [comment]
@lmiller1990 I refreshed the PR to update vue3-jest to jest 27.x.
If the changes look good, we could release an alpha, and try it out. As there are quite a few breaking changes, we could see some regressions.
If everything is fine, we can port the changes to vue2-jest (should be fairly straightforward).

**Comment by lmiller1990** [comment]
I did `yarn publish --tag 27.0.0-alpha.1` but I cannot see it on `npm`. Am I missing something? 🤔

**Comment by cexbrayat** [comment]
I gave it a spin on the unit tests of my Vue 3 ebook and everything looks fine \o/. I'll do more proper testing in the next few days, thanks for the release @lmiller1990

**Comment by lmiller1990** [comment]
Cool thanks, I also gave it a (very light) test but seems okay. This migration is actually pretty low risk, since we are publishing to a new package name.

I will try to test a bit more and upgrade my projects to Jest 27... lmk once you are feeling confident, and we can release the rest of the versions and merge it up. 

The main challenge will be messaging and communicating to our users on the versioning moving forward.

**Comment by nickmccurdy** [comment]
Are there any plans to support Vue 2 with Jest 27? This would be helpful for [Vue Testing Library](https://github.com/testing-library/vue-testing-library) because its default branch is targeting Vue 2, but I'd like it to support Jest 27 before we finish work on upgrading to Vue 3.

**Comment by cexbrayat** [comment]
@nickmccurdy It should be fairly straightforward to support it for Vue 2 once we're sure it works for Vue 3 👍

**Comment by cexbrayat** [comment]
I found a regression in TS projects with no Babel config, so I'll need to investigate and fix that.

**Comment by lmiller1990** [comment]
Ok, thanks. Happy to take this one a bit slow and make sure it's a smooth roll out.

**Comment by cexbrayat** [comment]
Update: this should now work properly for pure TS projects.

I opened another PR #366 to reflect the most common use-case for TS users that have no babel config (and this one is now on top of this extra test to make sure everything works).
I think it was working a bit by accident for jest v26...
I also added another commit, to replace the logic in the typescript transformer that was using babel to change the module to commonjs. I think it would be simpler to directly update the tsconfig we load with this module format? Please take a look and let me know.

It would be awesome to have an alpha.2 release if that looks good @lmiller1990

**Comment by florianrusch** [comment]
Hi all together,

I've tried today to bring my current project to jest 27, but I got this error if I run the tests:

```plain
$ npx jest --coverage

TypeError: Jest: a transform must export something.
    at .../node_modules/@jest/transform/build/ScriptTransformer.js:386:19
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Promise.all (index 3)
    at async ScriptTransformer.loadTransformers (.../node_modules/@jest/transform/build/ScriptTransformer.js:378:5)
    at async createScriptTransformer (.../node_modules/@jest/transform/build/ScriptTransformer.js:1106:3)
    at async .../node_modules/@jest/core/build/TestScheduler.js:269:31
    at async Promise.all (index 0)
    at async TestScheduler.scheduleTests (.../node_modules/@jest/core/build/TestScheduler.js:262:5)
    at async runJest (.../node_modules/@jest/core/build/runJest.js:387:19)
    at async _run10000 (.../node_modules/@jest/core/build/cli/index.js:408:7)
```

My setup:

package.json:
```json
{
  ...,
  "dependencies": {
    "vue": "^3.0.0",
    "vue-i18n": "^9.1.6",
    "vue-router": "^4.0.0-0",
    "vuex": "^4.0.0",
    "vuex-oidc": "^3.10.2",
    ...
  },
  "devDependencies": {
    "@types/jest": "^26.0.24",
    "@typescript-eslint/eslint-plugin": "^4.18.0",
    "@typescript-eslint/parser": "^4.18.0",
    "@vue/cli-plugin-babel": "~4.5.0",
    "@vue/cli-plugin-eslint": "~4.5.0",
    "@vue/cli-plugin-router": "~4.5.0",
    "@vue/cli-plugin-typescript": "~4.5.0",
    "@vue/cli-plugin-unit-jest": "~4.5.0",
    "@vue/cli-service": "~4.5.0",
    "@vue/compiler-sfc": "^3.0.0",
    "@vue/eslint-config-prettier": "^6.0.0",
    "@vue/eslint-config-typescript": "^7.0.0",
    "@vue/test-utils": "^2.0.0-rc.10",
    "eslint": "^6.7.2",
    "eslint-plugin-prettier": "^3.3.1",
    "eslint-plugin-vue": "^7.0.0",
    "jest": "^27.0.6",
    "jest-junit-reporter": "^1.1.0",
    "prettier": "^2.2.1",
    "sass": "^1.26.5",
    "sass-loader": "^8.0.2",
    "ts-jest": "^27.0.3",
    "typescript": "~4.1.5",
    "vue3-jest": "^27.0.0-alpha.1"
  }
}
```

jest.config.js:
```javascript
module.exports = {
  preset: '@vue/cli-plugin-unit-jest/presets/typescript-and-babel',
  transform: {
    '^.+\\.vue$': 'vue3-jest',
  },
  collectCoverageFrom: ['<rootDir>/src/**/*.{ts,vue,js}'],
  coverageDirectory: '<rootDir>/test-coverage',
  coverageReporters: ['text', 'cobertura'],
  testMatch: ['<rootDir>/src/**/*.spec.ts'],
  modulePaths: ['<rootDir>'],
  testResultsProcessor: './node_modules/jest-junit-reporter',
}
```

A quick google search brought me to this issue in the ts-jest project: kulshekhar/ts-jest#2612.
However the offered fix with clearing the jest cache didn't work for me.

Does anyone have a clue what is going wrong?

**Comment by cexbrayat** [comment]
Hi @florianrusch 
Thanks for giving it a try.
For the moment the CLI depends on jest 26, so you have to use Yarn and add a resolution in your package.json:
```json
"resolutions": {
  "jest": "27.0.6"
}
```

Run `yarn` to install the dependencies, and the error should go away, as jest 27 will be used.

Then you need to update your jest config, as the CLI preset will not work (it uses `vue-jest`).
Add something like:
```js
  transform: {
    '^.+\\.vue$': 'vue3-jest',
    '^.+\\.tsx?$': 'ts-jest',
    '^.+\\.jsx?$': 'babel-jest'
  },
  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/src/$1'
  },
```
to mimic what the preset does.

Then `yarn test:unit` should work.
vue3-jest@27.0.0-alpha.1 has an issue though, but as it looks like you use babel as well, it should be fine.

I'm interested to hear if vue3-jest works for you. WHen we're confident enough, I'll open a PR on the CLI to avoid all this manual work 😉

**Comment by florianrusch** [comment]
Thanks @cexbrayat for you reply. I directly tested it.

The following makes the error disappear:
```javascript
  transform: {
    '^.+\\.vue$': 'vue3-jest',
    '^.+\\.tsx?$': 'ts-jest',
    '^.+\\.jsx?$': 'babel-jest'
  },
```

But after I applied it (also tested with the other changes), I've got a new error:
```shell
$ npx jest --coverage
Test environment found at "/Users/fr/Workspace/ZF/web-app/node_modules/jest-environment-jsdom-fifteen/lib/index.js" does not export a "getVmContext" method, which is mandatory from Jest 27. This method is a replacement for "runScript".
Test environment found at "/Users/fr/Workspace/ZF/web-app/node_modules/jest-environment-jsdom-fifteen/lib/index.js" does not export a "getVmContext" method, which is mandatory from Jest 27. This method is a replacement for "runScript".

...

 FAIL  src/components/Header.spec.ts
  ● Test suite failed to run

    Jest worker encountered 4 child process exceptions, exceeding retry limit

      at ChildProcessWorker.initialize (node_modules/jest-worker/build/workers/ChildProcessWorker.js:193:21)
```

The first message seams to appear repeated for all tests.

**Comment by cexbrayat** [comment]
Try to add
```
testEnvironment: 'jsdom',
```
to the jest config maybe?

**Comment by florianrusch** [comment]
Awesome 🙂 It works! Thank you very much for your help!

To wrap up, I had to add the following lines into my jest config:
```javascript
  transform: {
    '^.+\\.vue$': 'vue3-jest',
    '^.+\\.tsx?$': 'ts-jest',
    '^.+\\.jsx?$': 'babel-jest',
  },
  testEnvironment: 'jsdom',
```

**Comment by lmiller1990** [comment]
`27.0.0-alpha.2` is now available.

**Comment by dergunovs** [comment]
Getting errors after update to vue3-jest@27.0.0-alpha.2
I tried to remove node_modules and reinstall the dependencies - no effect.

```
 FAIL  src/components/Example.spec.js        
  ● Test suite failed to run
    TypeError: Cannot read property 'cwd' of undefined
      at Object.getCacheKey (node_modules/babel-jest/build/index.js:297:33)
```

**Versions in my package.json:**
  ```
"dependencies": {
    "vue": "^3.1.5",
  },
  "devDependencies": {
    "@vue/cli-plugin-babel": "^4.5.13",
    "@vue/cli-plugin-unit-jest": "^4.5.13",
    "@vue/test-utils": "^2.0.0-rc.10",
    "babel-core": "^7.0.0-bridge.0",
    "babel-jest": "27.0.6",
    "jest": "27.0.6",
    "vue3-jest": "^27.0.0-alpha.2"
  },
  "resolutions": {
    "jest": "27.0.6"
  }
```

**Jest.config.js:**
```
module.exports = {
  moduleFileExtensions: ['js', 'vue'],
  transform: {
    '^.+\\.js$': 'babel-jest',
    '^.+\\.vue$': 'vue3-jest',
  },
  testEnvironment: 'jsdom',
  transformIgnorePatterns: ['/node_modules/'],
  moduleNameMapper: { '^@/(.*)$': '<rootDir>/src/$1' },
};
```

**Comment by cexbrayat** [comment]
Interesting. Is it still in error after `yarn test:unit --clear-cache`?

If so, it would be great if you can share a small repro in a github repo so we could investigate.

**Comment by lmiller1990** [comment]
I think the flag might be `--no-cache`.

**Comment by dergunovs** [comment]
I searched "jest" in package-lock.json. Found 24.x.x version in vue-cli dependencies. Manualy changed it to ^27. Then "npm i". Now everything is working.

**Comment by dergunovs** [comment]
Ok, it work's only 1 time. Second "npm i" returns jest 24 and errors.

**Comment by cexbrayat** [comment]
@dergunovs Yeah, it's easier to do with Yarn and its resolution feature. If the alpha phase goes smoothly, we'll update the CLI to support vue3-jest, and we'll be able to have jest 27 support without fussing around the dependency resolutions. I'd recommend using Yarn in the meantime if you want to keep using vue3-jest. Thanks for confirming this was not an issue with vue3-jest 👍

**Comment by cexbrayat** [comment]
FYI, I opened a feature request to the CLI team to include vue3-jest and jest 27 by default in a project https://github.com/vuejs/vue-cli/issues/6602

**Comment by lmiller1990** [comment]
Great!

I will review your other PRs in the next day or two. You should really have push access to this repo, you could ping Evan (I cannot give you access).

**Comment by tigohenryschultz** [comment]
> Getting errors after update to vue3-jest@27.0.0-alpha.2 I tried to remove node_modules and reinstall the dependencies - no effect.
> 
> ```
>  FAIL  src/components/Example.spec.js        
>   ● Test suite failed to run
>     TypeError: Cannot read property 'cwd' of undefined
>       at Object.getCacheKey (node_modules/babel-jest/build/index.js:297:33)
> ```
> 
> **Versions in my package.json:**
> 
> ```
> "dependencies": {
>   "vue": "^3.1.5",
> },
> "devDependencies": {
>   "@vue/cli-plugin-babel": "^4.5.13",
>   "@vue/cli-plugin-unit-jest": "^4.5.13",
>   "@vue/test-utils": "^2.0.0-rc.10",
>   "babel-core": "^7.0.0-bridge.0",
>   "babel-jest": "27.0.6",
>   "jest": "27.0.6",
>   "vue3-jest": "^27.0.0-alpha.2"
> },
> "resolutions": {
>   "jest": "27.0.6"
> }
> ```
> 
> **Jest.config.js:**
> 
> ```
> module.exports = {
>   moduleFileExtensions: ['js', 'vue'],
>   transform: {
>     '^.+\\.js$': 'babel-jest',
>     '^.+\\.vue$': 'vue3-jest',
>   },
>   testEnvironment: 'jsdom',
>   transformIgnorePatterns: ['/node_modules/'],
>   moduleNameMapper: { '^@/(.*)$': '<rootDir>/src/$1' },
> };
> ```

Was this ever resolved?

**Comment by lmiller1990** [comment]
No idea, a repo showing the issue would be helpful.

**Pull request merged**